core/src/main/java/org/apache/iceberg/actions/SizeBasedFileRewritePlanner.java

```diff
@@ -0,0 +1,361 @@
+   * that group is less than the target file size. This can also be thought of as the maximum number
+   * of wrongly sized files that could remain in a partition after rewriting.
+   */
+  public static final String MIN_INPUT_FILES = "min-input-files";
+
+  public static final int MIN_INPUT_FILES_DEFAULT = 5;
+
+  /** Overrides other options and forces rewriting of all provided files. */
+  public static final String REWRITE_ALL = "rewrite-all";
+
+  public static final boolean REWRITE_ALL_DEFAULT = false;
+
+  /**
+   * This option controls the largest amount of data that should be rewritten in a single file
+   * group. It helps with breaking down the rewriting of very large partitions which may not be
+   * rewritable otherwise due to the resource constraints of the cluster. For example, a sort-based
+   * rewrite may not scale to TB-sized partitions, and those partitions need to be worked on in
+   * small subsections to avoid exhaustion of resources.
+   */
+  public static final String MAX_FILE_GROUP_SIZE_BYTES = "max-file-group-size-bytes";
+
+  public static final long MAX_FILE_GROUP_SIZE_BYTES_DEFAULT = 100L * 1024 * 1024 * 1024; // 100 GB
+
+  private static final long SPLIT_OVERHEAD = 5L * 1024;
+
+  private final Table table;
+  private long targetFileSize;
+  private long minFileSize;
+  private long maxFileSize;
+  private int minInputFiles;
+  private boolean rewriteAll;
+  private long maxGroupSize;
+  private int outputSpecId;
+
+  protected SizeBasedFileRewritePlanner(Table table) {
+    this.table = table;
+  }
+
+  /** Expected target file size before configuration. */
+  protected abstract long defaultTargetFileSize();
+
+  /** Additional filter for tasks before grouping. */
+  protected abstract Iterable<T> filterFiles(Iterable<T> tasks);
+
+  /** Additional filter for groups. */
+  protected abstract Iterable<List<T>> filterFileGroups(List<List<T>> groups);
+
+  @Override
+  public Set<String> validOptions() {
+    return ImmutableSet.of(
+        TARGET_FILE_SIZE_BYTES,
+        MIN_FILE_SIZE_BYTES,
+        MAX_FILE_SIZE_BYTES,
+        MIN_INPUT_FILES,
+        REWRITE_ALL,
+        MAX_FILE_GROUP_SIZE_BYTES);
+  }
+
+  @Override
+  public void init(Map<String, String> options) {
+    Map<String, Long> sizeThresholds = sizeThresholds(options);
+    this.targetFileSize = sizeThresholds.get(TARGET_FILE_SIZE_BYTES);
+    this.minFileSize = sizeThresholds.get(MIN_FILE_SIZE_BYTES);
+    this.maxFileSize = sizeThresholds.get(MAX_FILE_SIZE_BYTES);
+    this.minInputFiles = minInputFiles(options);
+    this.rewriteAll = rewriteAll(options);
+    this.maxGroupSize = maxGroupSize(options);
+    this.outputSpecId = outputSpecId(options);
+
+    if (rewriteAll) {
+      LOG.info("Configured to rewrite all provided files in table {}", table.name());
+    }
+  }
+
+  protected Table table() {
+    return table;
+  }
+
+  protected boolean wronglySized(T task) {
```

Comment:
I know the method name is copied from existing code. I don't know if we are entertaining maybe some renaming to make it easier to read.

e.g., this one can be called `outsideDesiredFileSizeRange`.

Reply:
Renamed